src/java.base/share/classes/java/util/JapaneseImperialCalendar.java

```diff
@@ -1412,79 +1383,71 @@ public int getActualMaximum(int field) {
+                    CalendarDate d = gcal.newCalendarDate(TimeZone.NO_TIMEZONE);
+                    d.setDate(date.getNormalizedYear(), BaseCalendar.JANUARY, 1);
+                    int dayOfWeek = gcal.getDayOfWeek(d);
+                    // Normalize the day of week with the firstDayOfWeek value
+                    dayOfWeek -= getFirstDayOfWeek();
+                    if (dayOfWeek < 0) {
+                        dayOfWeek += 7;
+                    }
+                    int magic = dayOfWeek + getMinimalDaysInFirstWeek() - 1;
+                    if ((magic == 6) ||
+                        (date.isLeapYear() && (magic == 5 || magic == 12))) {
+                        yield 53;
+                    }
+                    yield 52;
                 }
 
                 if (jc == this) {
                     jc = (JapaneseImperialCalendar) jc.clone();
                 }
                 int max = getActualMaximum(DAY_OF_YEAR);
                 jc.set(DAY_OF_YEAR, max);
-                value = jc.get(WEEK_OF_YEAR);
-                if (value == 1 && max > 7) {
+                int weekOfYear = jc.get(WEEK_OF_YEAR);
+                if (weekOfYear == 1 && max > 7) {
                     jc.add(WEEK_OF_YEAR, -1);
-                    value = jc.get(WEEK_OF_YEAR);
+                    weekOfYear = jc.get(WEEK_OF_YEAR);
                 }
+                yield weekOfYear;
             }
-            break;
-
-        case WEEK_OF_MONTH:
-            {
-                LocalGregorianCalendar.Date jd = jcal.getCalendarDate(Long.MAX_VALUE,
-                                                                      getZone());
-                if (!(date.getEra() == jd.getEra() && date.getYear() == jd.getYear())) {
-                    CalendarDate d = gcal.newCalendarDate(TimeZone.NO_TIMEZONE);
-                    d.setDate(date.getNormalizedYear(), date.getMonth(), 1);
-                    int dayOfWeek = gcal.getDayOfWeek(d);
-                    int monthLength = gcal.getMonthLength(d);
-                    dayOfWeek -= getFirstDayOfWeek();
-                    if (dayOfWeek < 0) {
-                        dayOfWeek += 7;
-                    }
-                    int nDaysFirstWeek = 7 - dayOfWeek; // # of days in the first week
-                    value = 3;
-                    if (nDaysFirstWeek >= getMinimalDaysInFirstWeek()) {
-                        value++;
-                    }
-                    monthLength -= nDaysFirstWeek + 7 * 3;
-                    if (monthLength > 0) {
-                        value++;
-                        if (monthLength > 7) {
-                            value++;
-                        }
-                    }
-                } else {
+            case WEEK_OF_MONTH -> {
+                LocalGregorianCalendar.Date jd = jcal.getCalendarDate(Long.MAX_VALUE, getZone());
+                if (date.getEra() == jd.getEra() && date.getYear() == jd.getYear()) {
```

Comment:
I inverted the `if` condition here, as it was negated and a much shorter branch was in `else`.